lib/wallet-deposit/spec/customer-deposit-wallet.lagda.md

```diff
@@ -0,0 +1,391 @@
+
+### Tracking incoming funds
+
+Beyond assigning an address to a customer,
+the new wallet state returned by `createAddress`
+also tracks this address whenever new blocks are incorporated into the wallet state.
+For this purpose of tracking, we introduce an operation
+
+    getCustomerHistory : WalletState → Customer → List TxSummary
+
+which returns a list of transaction summaries. For a given transaction, such a summary reports the total `Value` spend or received at a specific address.
+
+    record ValueTransfer : Set where
+      field
+        spent    : Value
+        received : Value
+
+    open ValueTransfer
+
+    TxSummary : Set
+    TxSummary = Slot × TxId × ValueTransfer
+
+Note that `Value` includes both native coins (ADA) and
+user-defined assets, such as stablecoins NFTs.
+Also note that the customer deposit wallet does not support
+delegation and reward accounts, and the `spent` value
+can only be spent from transaction outputs.
+
+The function `getCustomerHistory` allows users to detect incoming
+transfers by observing the `received` value.
+
+The behavior of this function is best specified in terms of a function
+
+```agda
+      summarize : WalletState → Tx → List (Address × TxSummary)
+
+    getAddressSummary
+      : Address → List (Address × TxSummary) → List TxSummary
+    getAddressSummary address =
+      map snd ∘ filter (λ x → fst x == address)
+```
+
+which summarizes a single transaction. Specifically, the result of `getCustomerHistory` an aggregate of all previous transaction summaries.
+
+```agda
+    field
+      prop_getAddressHistory-summary
+        : ∀ (s : WalletState)
+            (c : Customer)
+            (address : Address)
+            (tx : Tx)
+        → (c , address) ∈ listCustomers s
+        → getCustomerHistory (applyTx tx s) c
+          ≡ (getAddressSummary address (summarize s tx))
+              ++ getCustomerHistory s c
+```
+
+Importantly, we only track an address if and only if it is a `knownCustomerAddress`.
+
+```agda
+      prop_tx-known-address
+        : ∀ (address : Address)
+            (s : WalletState)
+            (tx : Tx)
+        → (knownCustomerAddress address s ≡ True)
+        ⟷ (address ∈ map fst (summarize s tx))
+```
+
+### Creating transactions
+
+Finally, we expose an operation
+
+    createPayment : List (Address × Value) → WalletState → Maybe Tx
+
+which constructs and signs a transaction that sends given values to given addresses.
+
+First, this function will succeed in creating a transaction if there are sufficient
+funds available:
+
+```agda
+
+    field
+      totalValue : List (Address × Value) → Value
+      -- totalValue = mconcat ∘ map snd
+
+      maxFee : Value -- maximum fee of a transaction
+      exceeds : Value → Value → Set
+      _<>_ : Value → Value → Value
+
+      prop_createPayment-success
+        : ∀ (s : WalletState)
+            (destinations : List (Address × Value))
+        → exceeds (availableBalance s) (totalValue destinations <> maxFee)
```

Comment:
Ah, good point. 🤔 I was thinking about a "wallet-intrinsic" reason why the transaction might fail — not enough funds in `availableBalance s` — but it did not occur to me to check for reasons that are "destination-intrinsic", e.g. a `Value` in the `destinations` being lower than `minAda`.

Still, a statement of the form "If you have more than sufficient funds and your destinations are ok, it is guaranteed that you can make a transaction" would be highly desirable. 🤔🤑

Reply:
Alas, I'm not sure this is always possible. (Though I agree that such a property would be extremely desirable!)

Consider the following example:
- Alice has a wallet with a single UTxO `u` whose ada quantity `a` is the absolute minimum allowable for that UTxO.
- UTxO `u` may or may not have other non-ada assets. 
- Alice decides to pay the value of `u` minus `maxFee` to Bob.

In this very simple case, the requirement to pay a fee (for any non-zero fee value) will diminish the amount of ada available so that it's no longer possible to meet the minimum ada quantity requirement of the target UTxO (which will presumably also be `a`, barring a change to the protocol parameter that governs the minimum ada quantity).

In more complex scenarios, it's only possible to successfully make a payment if we allow ourselves to use more than one transaction to meet the goal. (Trivial example: the number of unique assets is so large that it would not be possible to write them all down in a single transaction.) This is what we do for the wallet balance migration algorithm, which creates as many transactions as are necessary in order to migrate the entire wallet balance.

Reply:
Johannes and I discussed a couple of ideas today.

First, let me state the goal: A **guarantee** on `createPayment` is a very strong statement, and in turn, it is perfectly fine for us to require **favorable** preconditions. In other words, we only guarantee to be able to make a transaction when the UTxO is very nice, and don't say much for the other cases. This is still a strict improvement over the status quo, where we have an algorithm that "works in practice, most of the time".

For example, in the draft specification, I have required that the `availableBalance` can cover the the maximum fee, even though `createPayment` will still be able to make transactions as the actual fee will be lower — but we don't guarantee it.

Then, part of the goal of a guarantee is to find favorable preconditions that are still useful.

One idea for preconditions is to classify the status of the UTxO into separate colors: `utxoStatus : UTxO → PParams → Color`

* `Green : Color` — `createPayment` is guaranteed to suceed.
* `Orange : Color` — `createPayment` may succeed, or may not. However, there is a separate function, `cleanupWallet : WalletState → Maybe Tx` that will create a transaction which moves the wallet status back to `Green`.
* `Red : Color` — you're on your own, sorry. Should still "work in practice, most of the time".

In other words, a `Green` wallet works always, an `Orange` wallet may not, but you will get a way of cleaning it up, and `Red` is what happens if you never clean up. In addition, we want a statement of the form that sending money to the wallet will not make the color worse — or we will have to start filtering terrible UTxO.

As both Johannes and you point out, the `Green` status may not be enough — the ability of making a transaction also depends on the size of the payment to be made. This means that one probably has to introduce an additional `Value` parameter (or result) in the `utxoStatus`.

Another idea, as you note, is to allow `createPayment` to return multiple transactions instead. (It may be more difficult to prove things about repeated coin selection, though.) In the end, the problem boils down to keeping the fee low.

In any case, for the purpose of the Customer Deposit Wallet, we currently don't need to offer or prove a guarantee — but I thought it would be worth writing something down in order to start the discussion. 🤓 

----

For reference, the difficult corner cases that we can currently think of

* `TxOut` so large that it takes almost all of the space of the transaction.
* `TxOut` too close to the minimum Ada quantity
* `TxIn` small compared to payment that we want to make, so that we would have to include too many of them. (Example: "Have 1 million x 1 ADA coins, want to send 1 x 1'000'000 ADA coin.")

Reply:
> We could make it part of the precondition: ", assuming all utxos are >= 1 ada, assuming <= 5 native assets per utxo, assuming Mainnet PParams, "

The difference would be that the assumption `utxos satisfy the minAda quantity => utxos contain more than ~1 ada` is unlikely to change drastically, and only through PParam updates / hard forks. An assumption like `<= 5 native assets per utxo` could be invalidated by one bad actor / customer.

To elaborate the example I gave yesterday: Suppose a user deposits as many SeaHorse NFTs as possible with the minimum amount of ada. This will be:

```
coinsPerUTxOByte * maxValSize = 4310  * 4000 lovelace ≈ 17 ada
```

The deposit wallet can either 
- deem the payment worthless as the 17 ada never can be separated from the NFTs.
- merge the NFTs with existing pure-ada utxos. However, this would increase the fee of selecting inputs in the future by `4000*44 = 0.176` ada, as ~4000 bytes of sea horse NFT change outputs would needed ...and make coin-selection more difficult due to the `txMaxSize` constraint as well.

It seems to me like this could be a serious problem 🤔 

So we could have a way to ignore payments containing native tokens.  Or one could perhaps also imagine subtracting the minAda from the ada value of ada+token and tracking them in a separate `locked ada` field (at least in the tx history of a normal wallet).

Reply:
> An assumption like <= 5 native assets per utxo could be invalidated by one bad actor / customer.

That's true, but I suppose the best way to deal with this is to not count it towards the `availableBalance` — we can always ignore UTxOs which are too bad. 🤔 I.e. the precondition can apply a filter `utxoThatCounts = { u ∈ utxo |  u has <= 5 native assets } /\ availableBalance utxoThatCounts >= …`.

> . Or one could perhaps also imagine subtracting the minAda from the ada value of ada+token and tracking them in a separate locked ada field (at least in the tx history of a normal wallet).

I agree that there appears to be an "unspendable ADA" part associated with each native asset, but I'm not sure what that is. 🤔

I believe that each asset is bounded by ~70 bytes: A `policy_id` is 28 bytes, an `asset_name` is 32 bytes (give or take a CBOR tag), and then 8 bytes (?) for the `uint` representing the value. Regardless of whether you group this with other assets, or whether you keep a single output with this asset, you need to pair it with `4310 * 70 = 0.3` ADA in order to have a valid output — I think that this amount qualifies as *unspendable*, and we can subtract it from the balance.

But I believe that it's — as soon as we account for these `0.3` ADA, any excess ADA value can be given to other parties and/or used to cover fees.

(Just to clarify: `maxValSize = 5000` (Bytes) according to 
https://book.world.dev.cardano.org/environments/mainnet/alonzo-genesis.json .)

Reply:
In analogy to "free electron" and "bound electron" in atomic physics, one could call the ~0.3 ADA "bound ADA", because they are bound to a native asset, and the difference "free ADA", because those are ADA that can freely leave the wallet. 🤔 🤓

Reply:
^ two values combing to form a more ada-efficient value

Reply:
My bet would be that we should just have a whitelist of recognised tokens though 🤔

Reply:
A simple solution to begin with would be to hardcode an empty whitelist (i.e. only discover pure-ada utxos).

Reply:
> Although, right, if you do configure it to allow some non-ada tokens, then we do run into the problem of "bound ada" affecting or needing to affect the accounting...

Yes. 🤔

I suppose that the most "versatile" UTxO distribution is the following:

* 1 UTxO for each `(token_policy, asset_name)` pair, containing both all of the tokens of this pair, as well as ~0.3 ADA.
* 1 UTxO containing the remaining ADA.

This minimizes the bound ADA, and maximizes flexibility for payment. Essentially, each UTxO acts like a single "account" for each currency.